### PR TITLE
Complete support of bk:space and bk:space+dots gestures for Seika Notetaker

### LINKED ISSUE #16642: The input translation table option doesn't work for some braille displays
### Steps to reproduce:

Case one:

1. Open the braille panel of the preference settings.
2. Set the input translation table to Unicode.
3. Press the braille space key in an editable control.

Case two:

1. Press `NVDA+1`, "input help on".
2. Press `space + dot 1` on the braille keyboard of Seika NoteTaker.

### Actual behavior:

In the case one, an ordinary space character appears. In the case two, NVDA says "dot 1".

### Expected behavior:

In the case one, a braille space (`⠀`) should appear. In the case two, NVDA should say "space with dot 1".

### NVDA logs, crash dumps and other attachments:

None. This is an unreasonable behavior due to the implementation, rather than an exception.

#### NVDA installed/portable/running from source:

Portable.

#### NVDA version:

2024.1

#### Windows version:

Windows 10 22H2 (10.0.19045) workstation AMD64

#### Name and version of other software in use when reproducing the issue:

None.

#### Other information about your system:

None.

#### Does the issue still occur after restarting your computer?

Yes.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

No.

#### If NVDA add-ons are disabled, is your problem still occurring?

Yes.

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?

Yes.

### PR DESCRIPTION
**Pull request by school510587**
### Link to issue number:

Closes [#16642](https://github.com/nvaccess/nvda/issues/16642).

### Summary of the issue:

Two drivers are mentioned in the issue, eurobraille and seikantk. The user can change an option of the former to solve the issue. However, there is no feasible approach for the user to solve the issue for the latter manually, which is the aim of the PR.

### Description of user facing changes

The user can perceive the changes by three ways.

1. Observe the input help. On pressing dots and/or space/backspace, names of braille input gestures are reported correctly.
2. Run braille input translation. The result must be as expected, while the old seikantk driver behaves incorrectly.
3. Use add-ons relying on braille input. In fact, this is the main demand that users reqest for the modification. Correct braille input leads to correct result.

### Description of development approach

4 cases of input gestures are explicitly specified in seikantk.py, and the algorithm description is available in the commit log.

### Testing strategy:

The most simple way to test is to run input help. Be sure that your working braille display is Seika Notetaker.

Case 1:

1. Press any dot along with either space or backspace on the braille display.
2. Expect "space with dot XXX" to be announced.

Case 2:

1. Press either space or backspace on the braille display.
2. Expect "Inputs braille dots via the braille keyboard" to be announced.

It is OK to test the change through the input translation table.

1. Set the input translation table to Unicode.
2. Press either space or backspace on the braille display.
3. A braille space is generated, rather than an ordinary space character.

### Known issues with pull request:

There are two notable issues currently.

1. Entries in gestureMap for br(seikantk):d7 and br(seikantk):d8 may be redundant.
2. With InputGestureRouting, the routing parameter of InputGesture.__init__ seens useless.
3. Incorrect internal br(seikantk) ID for some bk:space+dots gestures.

### Code Review Checklist:

- [ ] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [ ] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] API is compatible with existing add-ons.
- [ ] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Enhanced input handling in braille displays by combining `keys` and `dots` into a single gesture, improving the user experience.
- **Bug Fixes**
  - Corrected the input gesture logic to ensure `space` key mappings are no longer included erroneously, preventing conflicts.
- **Improvements**
  - Updated default settings for input gestures, ensuring smoother operation and better default handling for users.

### CODE REVIEW

**Comment by school510587** [comment]
Hi @md_curran,

Sorry, I am not familiar to the lint check. I think I got nothing from the AppVeyorBot report. If I must do anymore before merge, please tell me. The PR is important for many Taiwanese users, because it determines whether [BrlIMEHelper](https://github.com/school510587/BrlIMEHelper/) works for Seika Notetaker.

BrlIMEHelper is an add-on converting braille input into IME emulation, which can be thought of as another implementation of braille input handler. Seika Notetaker, called BPDA (亮點) by Taiwanese, is a majority of braille displays that students can borrow from the government organization. Thus, students with Seika Notetakers are eager for fix of the driver to facilitate their learning.

Thanks.

**Comment by hwf1324** [comment]
> Hi @md_curran,
> 
> Sorry, I am not familiar to the lint check. I think I got nothing from the AppVeyorBot report. If I must do anymore before merge, please tell me. The PR is important for many Taiwanese users, because it determines whether [BrlIMEHelper](https://github.com/school510587/BrlIMEHelper/) works for Seika Notetaker.
> 
> BrlIMEHelper is an add-on converting braille input into IME emulation, which can be thought of as another implementation of braille input handler. Seika Notetaker, called BPDA (亮點) by Taiwanese, is a majority of braille displays that students can borrow from the government organization. Thus, students with Seika Notetakers are eager for fix of the driver to facilitate their learning.
> 
> Thanks.

See: https://ci.appveyor.com/project/NVAccess/nvda/builds/50160022#L5371

**Comment by seanbudd** [comment]
I would encourage running `runlint.bat` before pushing any changes

**Comment by seanbudd** [comment]
Can you please add a [changelog entry](https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/contributing.md#change-log-entry) detailing this bug fix?

**Comment by school510587** [comment]
Hi,

I have uploaded the new changelog entry. However, the md file now uses different symbol `*` from the changelog entry specification `-`. I think it would be better to update the documentation.

File: user_docs/en/changes.md [hunk 1]
```diff
@@ -9,6 +9,7 @@
 ### Bug Fixes
 
 * NVDA once again relies on UIA events for caret movement in XAML and WPF text controls, rather than only on manual querying of the caret position. (#16817, @LeonarddeR)
+* The Seika Notetaker driver now correctly generates braille input for braille space and space with dots gestures. (#16642, @school510587)
```
Comment (seanbudd) [inline user_docs/en/changes.md:12]: Is this correct? I think there's a typo

```suggestion
* The Seika Notetaker driver now correctly generates braille input for space and backspace with dots gestures. (#16642, @school510587)
```

**Comment by seanbudd** [comment]
@school510587 good catch with the markdown issue. Do you mind updating that file in a separate PR?

File: user_docs/en/changes.md [hunk 2]
```diff
@@ -9,7 +9,7 @@
 ### Bug Fixes
 
 * NVDA once again relies on UIA events for caret movement in XAML and WPF text controls, rather than only on manual querying of the caret position. (#16817, @LeonarddeR)
-* The Seika Notetaker driver now correctly generates braille input for braille space and space with dots gestures. (#16642, @school510587)
+* The Seika Notetaker driver now correctly generates braille input for space, backspace and space/backspace with dots gestures. (#16642, @school510587)
```
Comment (seanbudd) [inline user_docs/en/changes.md:12]:
```suggestion
* The Seika Notetaker driver now correctly generates braille input for space, backspace and dots with space/backspace gestures. (#16642, @school510587)
```

**Comment by school510587** [comment]
Hi @seanbudd,

I think I should clearly specify which gestures are influenced, and there are actually 5: space, backspace, space with dots, backspace with dots, and space with backspace with dots. So I use "space, backspace and space/backspace with dots". Please tell me if there is any better representation, thanks.

**Review by seanbudd — APPROVED** [review]

**Comment by school510587** [comment]
> @school510587 good catch with the markdown issue. Do you mind updating that file in a separate PR?

OK. Please see [#16860](https://github.com/nvaccess/nvda/pull/16860).

**Pull request merged**